### ISSUE TITLE
Ability to add custom .ftl pages

### ISSUE THREAD

**Issue by asashay** [InseeFrLab/keycloakify#7]
Hello!

We have a highly customized Keycloak setup and have added a couple of additional '.ftl' pages to our login & signup flows.

As far as I know, the current version does not support such a thing as building arbitrary Freemarker pages. The only way I see to achieve this is to fork the library and add them manually to the build script.

Will it be possible to incorporate such functionality into the library? I could work on the PR if it's feasible (any advice is more than welcome ofc).

P.S. Thanks for the lib, it's a great piece of work, saved so much time for us 🙌

**Comment by garronej** [comment]
Hi,  
Thank you for you willingness to contribute. 👍🏻  
Is your project opensource? It will helm me understand the usecase if I could have a look on one of the custom page you made.

**Comment by asashay** [comment]
Hi @garronej,

thanks for the prompt response! 

It's not open-source. It's an app that allows to login and signup using biometric data. Thus we added a bunch of extra Freemarker templates to login theme to cover all our use-cases. E.g. login theme folder looks like - https://share.getcloudapp.com/yAurXePD. I believe we used a custom authenticator to achieve this. Not 100% sure though, cause I'm only doing front-end. 

The ideal result is to be able to build files matching ones from above using `keycloakify`.

Best,

**Comment by garronej** [comment]
OK,  
The screen shot you sent me must be ip protected because I get a 403 but whatever, I get the idea.  
Well, this it definitely something that Keycloakify should support.  

Today the only way to add support for a new page is [this](https://github.com/InseeFrLab/keycloakify#some-pages-still-have-the-default-theme-why)  

There is one main task that needs to be done if we want to really support addition of custom pages without forking. It's the way we convert the `.ftl` values into JavaScript objects ( how we build the `kcContext`).  
Currently what I am doing is  
1) I look at the `.ftl` files, [example with login-otp.ftl](https://github.com/InseeFrLab/keycloakify/releases/download/v0.0.1/login-otp.ftl), to see what global variable are used to construct the page. In the example I see that `otpLogin.userOtpCredentials` is used and that it seems to be an array of `{ id: string; userLabel: string }`.
2)  I write a ftl script to convert the `.ftl` values into a js object: [here with login-otp.ftl](https://github.com/InseeFrLab/keycloakify/blob/main/src/bin/build-keycloak-theme/generateFtl/login-otp.ftl)  

This approach is far from optimal because it will never let us support Keycloak plugins that introduce non-standard value like [keycloak-mail-whitelisting](https://github.com/micedre/keycloak-mail-whitelisting) that introduce `authorizedMailDomains`.  
I will also always require people to fork the project if they want to add support for other pages.  

What would be great would be, instead of having [one script for every page](https://github.com/InseeFrLab/keycloakify/tree/main/src/bin/build-keycloak-theme/generateFtl), to come up with a generic `ftl` script that lists all the global of the `data_model` and convert it into a JavaScript object.  
Something like that would be the starting point:  
```ftl
<#list .data_model?keys as var>
    ${var}
</#list>
```  

Long story short. If you want to implement support for the pages used at your company, the easiest way is to create your own fork and do as described [here](https://github.com/InseeFrLab/keycloakify#some-pages-still-have-the-default-theme-why).  
If, however, you are willing to contribute for the project PR are very welcome!
  
Kind regards,

**Comment by asashay** [comment]
Thanks for the hints!

Will definitely come back with PR if I find a way to add custom ftl pages to the lib elegantly.

Cheers!

**Comment by garronej** [comment]
Great thanks!

**Comment by garronej** [comment]
Hi @asashay, 
In v2 it is now possible to add custom `.ftl` pages

**Comment by asashay** [comment]
Hey, @garronej, that's huge!

And what's the integration drill? How are they supposed to be added? 

I've ended up creating all new `.ftl` pages, mocks, contexts... and having them sit in a fork.

Best,

**Comment by garronej** [comment]
If you want to go this way checkout the demo setup provided [here](https://github.com/garronej/keycloakify-demo-app/tree/look_and_feel)

Main takeaways are:
- You must declare your custom pages in the package.json. [example](https://github.com/garronej/keycloakify-demo-app/blob/4eb2a9f63e9823e653b2d439495bda55e5ecc134/package.json#L17-L22)
- (TS only) You must declare theses page in the type argument of the getter 
  function for the `kcContext` in order to have the correct typings. [example](https://github.com/garronej/keycloakify-demo-app/blob/4eb2a9f63e9823e653b2d439495bda55e5ecc134/src/KcApp/kcContext.ts#L16-L21)
- (TS only) If you use Keycloak plugins that defines non standard `.ftl` values
  (Like for example [this plugin](https://github.com/micedre/keycloak-mail-whitelisting) 
  that define `authorizedMailDomains` in `register.ftl`) you should 
  declare theses value to get the type. [example](https://github.com/garronej/keycloakify-demo-app/blob/4eb2a9f63e9823e653b2d439495bda55e5ecc134/src/KcApp/kcContext.ts#L6-L13)
- You should provide sample data for all the non standard value if you want to be able
  to debug the page outside of keycloak. [example](https://github.com/garronej/keycloakify-demo-app/blob/4eb2a9f63e9823e653b2d439495bda55e5ecc134/src/KcApp/kcContext.ts#L28-L43)

**Comment by asashay** [comment]
Thanks for the details!

I failed to set it up in the new React (TS) app for some reason 🤔. But keycloakify-demo-app builds fine, will dig some more.

Is there a way to load custom pages with mocked context btw?

**Comment by garronej** [comment]
Let me know if you need help. 
I'll be happy to help you set it up even via remote session.  
Yes I have done an extensive work so that you can fully customise the mocked context. 
You can overwrite the values provided by default and set values for non standard `ftl` variables.  
Check [this](https://github.com/garronej/keycloakify-demo-app/blob/4eb2a9f63e9823e653b2d439495bda55e5ecc134/src/KcApp/kcContext.ts#L30-L31) or [this](https://github.com/InseeFrLab/onyxia-web/blob/024d61d3d97799cffd7f25930729cc6ba933ec22/src/app/components/KcApp/kcContext.ts#L21-L31) out.

**Comment by asashay** [comment]
Hey @garronej , indeed that's a big addition to v1 💪.

I'm still trying to make it work for our app and am having issues after creating custom `kcContext`

Tried it out with your example and got the same error. 

Basically, when I am addind custom value to `login.ftl` I can't build the app, producing an error:
```
Property 'authorizedMailDomains' does not exist on type 'KcContext_Register'.
  Property 'authorizedMailDomains' does not exist on type 'Register'. 
```
And I only tried to add custom property to `login.ftl` template 🤔.

I forked the repo to create a reproducible example.
Here [link](https://github.com/asashay/keycloakify-demo-app/blob/621c3e521ebc040c8c783014ae0b2ec8c727b296/src/KcApp/kcContext.ts#L22) to the file with changes. 

Any ideas why it's happening?

Best,

**Comment by garronej** [comment]
Sorry you had to go through this, this is definitely a bug on my side... 🤨  
Let me fix it and I'll come back to you asap

**Comment by garronej** [comment]
@asashay Sorry again for this bug. I implemented this feature in TDD but still my unit test let this huge bug goes noticed.  It should be fixed now.

**Comment by asashay** [comment]
hey @garronej no worries at all! Actually, now that it's fixed I was able to rewrite all the custom pages to use `keycloakify` v2 .

I think it's a great addition to the lib and covers all our needs so far. Thanks for adding this feature!
